src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs

```diff
@@ -893,53 +901,43 @@ private bool ConsumeStringNextSegment()
                                                 byte nextByte = localBuffer[j];
                                                 if (!JsonReaderHelper.IsHexDigit(nextByte))
                                                 {
-                                                    _currentPosition = startPosition;
+                                                    RollBackStateWhenParsingString(rollBackState, isError: true);
                                                     ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidHexCharacterWithinString, nextByte);
                                                 }
-                                                if (j - idx > numberOfHexDigits)
+                                                numberOfHexDigits++;
+                                                _bytePositionInLine++;
+                                                if (numberOfHexDigits >= 4)
                                                 {
-                                                    if (movedToNext)
-                                                    {
-                                                        nextCharEscaped = false;
-                                                        goto StartOfLoop;
-                                                    }
-                                                    goto DoneWithHex;
+                                                    nextCharEscaped = false;
+                                                    idx = j + 1; // Skip the 4 hex digits, the for loop accounts for idx incrementing past the 'u'
```

Comment:
This was a valid product bug, we weren't updating idx (resulting in extra iterations).

Reply:
Oh? Looks like it used to be handled with `goto DoneWIthHex`?

Reply:
Not always. In the case when data straddles a segment and `moveNext` was true, we would instead `goto StartOfLoop` without idx being updated. That has been fixed here.